kubewatcher/kubewatcher.go

```diff
@@ -258,50 +257,71 @@ func getResourceVersion(obj runtime.Object) (string, error) {
 func (ws *watchSubscription) eventDispatchLoop() {
 	log.Println("Listening to watch ", ws.Watch.Metadata.Name)
 	for {
-		for {
-			ev, more := <-ws.kubeWatch.ResultChan()
-			if !more {
+		// check watchSubscription is stopped or not before waiting for event
+		// comes from the kubeWatch.ResultChan(). This fix the edge case that
+		// new kubewatch is created in the restartWatch() while the old kubewatch
+		// is being used in watchSubscription.stop().
+		if ws.isStopped() {
+			ws.kubeWatch.Stop()
+			break
+		}
+		ev, more := <-ws.kubeWatch.ResultChan()
+		if !more {
+			if ws.isStopped() {
+				// watch is removed by user.
 				log.Println("Watch stopped", ws.Watch.Metadata.Name)
-				break
-			}
-			if ev.Type == watch.Error {
-				e := apierrs.FromObject(ev.Object)
-				log.Println("Watch error, retrying in a second: %v", e)
-				// Start from the beginning to get around "too old resource version"
-				ws.lastResourceVersion = ""
-				time.Sleep(time.Second)
-				break
-			}
-			rv, err := getResourceVersion(ev.Object)
-			if err != nil {
-				log.Printf("Error getting resourceVersion from object: %v", err)
+				return
 			} else {
-				log.Printf("rv=%v", rv)
-				ws.lastResourceVersion = rv
+				// watch closed due to unknown issues, restart it.
+				log.Println("Watch stopped unexpectedly", ws.Watch.Metadata.Name)
+				err := ws.restartWatch()
+				if err != nil {
+					log.Panicf("Failed to restart watch: %v", err)
+				}
+				continue
 			}
+		}
 
-			// Serialize the object
-			var buf bytes.Buffer
-			err = printKubernetesObject(ev.Object, &buf)
-			if err != nil {
-				log.Printf("Failed to serialize object: %v", err)
-				// TODO send a POST request indicating error
-			}
+		if ev.Type == watch.Error {
+			e := apierrs.FromObject(ev.Object)
+			log.Println("Watch error, retrying in a second: %v", e)
+			// Start from the beginning to get around "too old resource version"
+			ws.lastResourceVersion = ""
+			time.Sleep(time.Second)
+			continue
```

Comment:
Hmm, the old code restarts the watch when this happens, but the new code tries to read from the watch again. I think you need to call `ws.restartWatch()` here.